runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.interceptors
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.hashing.*
+import aws.smithy.kotlin.runtime.http.*
+import aws.smithy.kotlin.runtime.http.operation.*
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.request.header
+import aws.smithy.kotlin.runtime.http.request.toBuilder
+import aws.smithy.kotlin.runtime.util.*
+import kotlinx.coroutines.*
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Mutate a request to enable flexible checksums.
+ *
+ * If the checksum will be sent as a header, calculate the checksum.
+ *
+ * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
+ * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
+ * after the entire body has been streamed.
+ *
+ * @param checksumAlgorithmNameInitializer a function which parses the input [I] to return the checksum algorithm name
+ */
+@InternalApi
+public class FlexibleChecksumsRequestInterceptor<I>(
+    private val checksumAlgorithmNameInitializer: (I) -> String?,
+) : HttpInterceptor {
+    private var checksumAlgorithmName = CompletableDeferred<String>()
+
+    override fun readAfterSerialization(context: ProtocolRequestInterceptorContext<Any, HttpRequest>) {
+        @Suppress("UNCHECKED_CAST")
+        val input = context.request as I
+        checksumAlgorithmNameInitializer(input)?.let { checksumAlgorithmName.complete(it) }
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.getLogger<FlexibleChecksumsRequestInterceptor<I>>()
+        if (!checksumAlgorithmName.isCompleted) {
+            logger.debug { "no checksum algorithm specified, skipping flexible checksums processing" }
+            return context.protocolRequest
+        }
+
+        val req = context.protocolRequest.toBuilder()
+
+        check(context.protocolRequest.body.contentLength != null && context.protocolRequest.body.contentLength!! > 0) {
```

Comment:
correctness: `-1` indicates an unknown body length (i.e. transfer encoding chunked). Is this correct given that?

Reply:
Yeah, this is correct because the `chunked` `Transfer-Encoding` gets set during signing, which is downstream of this check